src/main/java/com/commercetools/sync/producttypes/utils/ProductTypeUpdatePlainEnumActionUtils.java

```diff
@@ -0,0 +1,136 @@
+        }
+    }
+
+
+    /**
+     * Compares a list of old {@link EnumValue}s with a list of new {@link EnumValue}s for a given attribute
+     * definition.
+     * The method serves as a generic implementation for plain enum values syncing. The method takes in functions
+     * for building the required update actions (AddEnumValue, RemoveEnumValue, ChangeEnumValueOrder and 1-1
+     * update actions on plain enum values (e.g. changeLabel) for the required resource.
+     *
+     * @param attributeDefinitionName the attribute name whose plain enum values are going to be synced.
+     * @param oldEnumValues           the old list of plain enum values.
+     * @param newEnumValues           the new list of plain enum values.
+     * @return a list of plain enum values update actions if the list of plain enum values is not identical.
+     *         Otherwise, if the plain enum values are identical, an empty list is returned.
+     * @throws DuplicateKeyException in case there are plain enum values with duplicate keys.
+     */
+    @Nonnull
+    private static List<UpdateAction<ProductType>> buildUpdateActions(
+        @Nonnull final String attributeDefinitionName,
+        @Nonnull final List<EnumValue> oldEnumValues,
+        @Nonnull final List<EnumValue> newEnumValues)
+        throws DuplicateKeyException {
+
+        final List<UpdateAction<ProductType>> removeEnumValuesUpdateActions = buildRemoveEnumValuesUpdateActions(
+            attributeDefinitionName,
+            oldEnumValues,
+            newEnumValues
+        )
+            .map(Collections::singletonList)
+            .orElse(emptyList());
+
+        final List<UpdateAction<ProductType>> matchingEnumValuesUpdateActions =
+            buildMatchingEnumValuesUpdateActions(
+                attributeDefinitionName,
+                oldEnumValues,
+                newEnumValues,
+                PlainEnumUpdateActionsUtils::buildActions
+            );
+
+        final List<UpdateAction<ProductType>> addEnumValuesUpdateActions = buildAddEnumValuesUpdateActions(
+            attributeDefinitionName,
+            oldEnumValues,
+            newEnumValues,
+            AddEnumValue::of
+        );
+
+        final List<UpdateAction<ProductType>> changeEnumValuesOrderUpdateActions =
+            buildChangeEnumValuesOrderUpdateAction(
+                attributeDefinitionName,
+                oldEnumValues,
+                newEnumValues,
+                ChangeEnumValueOrder::of
+            )
+                .map(Collections::singletonList)
+                .orElse(emptyList());
+
+        return Stream.concat(
+            Stream.concat(
+                removeEnumValuesUpdateActions.stream(),
+                matchingEnumValuesUpdateActions.stream()
+            ),
+            Stream.concat(
+                addEnumValuesUpdateActions.stream(),
+                changeEnumValuesOrderUpdateActions.stream()
+            )
+
+        ).collect(Collectors.toList());
+    }
```

Comment:
This method looks exactly the same as ProductTypeUpdateLocalizedEnumActionUtils#buildActions. Why don't you make a generic implementation and reuse. Like this:


````java

    @Nonnull
    private static <T extends WithKey> List<UpdateAction<ProductType>> buildUpdateActions(
        @Nonnull final String attributeDefinitionName,
        @Nonnull final List<T> oldEnumValues,
        @Nonnull final List<T> newEnumValues,
        @Nonnull final TriFunction<String, T, T, List<UpdateAction<ProductType>>> matchingEnumActionBuilder,
        @Nonnull final BiFunction<String, T, UpdateAction<ProductType>> addEnumActionBuilder,
        @Nonnull final BiFunction<String, List<T>, UpdateAction<ProductType>> changeOrderActionBuilder){

        final List<UpdateAction<ProductType>> removeEnumValuesUpdateActions = buildRemoveEnumValuesUpdateActions(
            attributeDefinitionName, oldEnumValues, newEnumValues)
            .map(Collections::singletonList)
            .orElse(emptyList());

        final List<UpdateAction<ProductType>> matchingEnumValuesUpdateActions =
            buildMatchingEnumValuesUpdateActions(attributeDefinitionName, oldEnumValues, newEnumValues,
                matchingEnumActionBuilder);

        final List<UpdateAction<ProductType>> addEnumValuesUpdateActions = buildAddEnumValuesUpdateActions(
            attributeDefinitionName, oldEnumValues, newEnumValues, addEnumActionBuilder);

        final List<UpdateAction<ProductType>> changeEnumValuesOrderUpdateActions =
            buildChangeEnumValuesOrderUpdateAction(attributeDefinitionName, oldEnumValues, newEnumValues,
                changeOrderActionBuilder)
                .map(Collections::singletonList)
                .orElse(emptyList());

        return Stream.concat(
            Stream.concat(removeEnumValuesUpdateActions.stream(), matchingEnumValuesUpdateActions.stream()),
            Stream.concat(addEnumValuesUpdateActions.stream(), changeEnumValuesOrderUpdateActions.stream()))
                     .collect(Collectors.toList());
    }
````


and then reuse it for both by doing:
for `ProductTypeUpdatePlainEnumActionUtils`
````java
buildUpdateActions(attributeDefinitionName, oldEnumValues, newEnumValues, PlainEnumUpdateActionsUtils::buildActions, AddEnumValue::of, ChangeEnumValueOrder::of);
````

and for ProductTypeUpdateLocalizedEnumActionUtils:
````java
buildUpdateActions(attributeDefinitionName, oldEnumValues, newEnumValues, LocalizedEnumUpdateActionsUtils::buildActions, AddLocalizedEnumValue::of, ChangeLocalizedEnumValueOrder::of);
````

Reply:
@heshamMassoud I tried your way first and I explained the reason of "copying" to Lam. This is my previous comment 


"The ProductTypeUpdateLocalizedEnumActionUtils and ProductTypeUpdatePlainEnumActionUtils classes are similar in the buildUpdateAction method and someone could argue that I could also reuse the code in the ProductTypeUpdateEnumActionsUtils class, the same as it's done with other methods.

I did it that way first but the code ended up more tight and difficult to understand since all three particular methods for adding, removing and change order of the enums have to be passed as parameters in the common method placed at ProductTypeUpdateEnumActionsUtils, making this method to have 6 parameters where some of them are BiFunction, TriFunction... (long and "complex" signature)

So I finally decided to repeat some code in ProductTypeUpdateLocalizedEnumActionUtils and ProductTypeUpdateLocalizedEnumActionUtils in order to clarify and decouple the buildUpdateAction method of both classes."

Reply:
I think it's ok to keep them copied also, I agree with your arguments. It was just an indicator for reuse right away and taking advantage of generics when they look the same. So we don't have to have the exact same code twice. But your arguments are also valid enough, It's fine for now, since it's only going to be reused by only these two.  From experience we had some generic methods which maybe had many params but were very helpful to use by any sync process implementation and saved a lot of time (and also the dev didn't need to know about how it's working inside), this was the [custom update action builder](https://github.com/commercetools/commercetools-sync-java/blob/master/src/main/java/com/commercetools/sync/commons/utils/CustomUpdateActionUtils.java#L125-L126) util method - it was used easily by almost every sync process implementation for example latley in the [price custom type sync](https://github.com/commercetools/commercetools-sync-java/blob/master/src/main/java/com/commercetools/sync/products/utils/ProductVariantPriceUpdateActionUtils.java#L116).

We can leave it like this for now, and see what happens in the future, if we see that we are changing always in both the same things, then we can generify the method.